src/vcf/fasta.cpp

```diff
@@ -0,0 +1,131 @@
+/**
+ * Copyright 2018 EMBL - European Bioinformatics Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <memory>
+
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/iostreams/copy.hpp>
+
+#include "bioio/bioio.hpp"
+
+#include "util/file_utils.hpp"
+#include "util/logger.hpp"
+#include "util/stream_utils.hpp"
+#include "util/string_utils.hpp"
+#include "vcf/compression.hpp"
+#include "vcf/fasta.hpp"
+#include "vcf/string_constants.hpp"
+
+std::shared_ptr<std::istream> create_gunzipped_stream(std::shared_ptr<std::istream> stream) {
+  std::shared_ptr<std::stringstream> ss(new std::stringstream());
+  try {
+    boost::iostreams::filtering_istream decompressedStream;
+    ebi::vcf::create_uncompressed_stream(*stream, ebi::vcf::GZ, decompressedStream);
+    boost::iostreams::copy(decompressedStream, *ss);
+  } catch (std::exception const &ex) {
+    BOOST_LOG_TRIVIAL(error) << ex.what();
+  }
+
+  return std::static_pointer_cast<std::istream>(ss);
+}
+
+std::shared_ptr<std::istream> create_decompressed_stream(std::shared_ptr<std::istream> stream) {
+  std::vector<char> line;
+  ebi::vcf::get_magic_num(*stream, line);
+  std::string file_extension = ebi::vcf::get_compression("", line);
+
+  if ( file_extension == ebi::vcf::GZ ) {
+    return create_gunzipped_stream(stream);
```

Comment:
can we use `ebi::vcf::create_uncompressed_stream` directly here ?

Reply:
In `create_gunzipped_stream` function it is calling `ebi::vcf::create_uncompressed_stream`. But for a `gz` stream, it needs to go futher on that and decompress it to allow random access.